lib/deimos/activerecord/batch_consumption.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Deimos
+  module ActiveRecord
+    # Batch consume methods
+    module BatchConsumption
+      # Handle a batch of Kafka messages
+      # @param payloads [Array<Hash>] Decoded payloads.
+      # @param metadata [Hash] Information about batch, including keys.
+      def consume_batch(payloads, metadata)
+        # Create [key, value] pairs, even if no keys are decoded
+        messages = payloads.
+          zip(metadata[:keys]).
+          map { |p, k| [k, p] }
+
+        Deimos.instrument('ar_consumer.consume_batch', topic: metadata[:topic], messages: messages) do
+          slices = _slice_batch(messages)
+
+          # The entire batch should be treated as one transaction so that if
+          # any message fails, the whole thing is rolled back
+          ::ActiveRecord::Base.transaction do
+            slices.each do |slice|
+              removed, upserted = slice.partition { |_, v| v.nil? }
+
+              upsert_records(upserted) unless upserted.empty?
+              remove_records(removed) unless removed.empty?
+            end
+          end
+        end
+      end
+
+    protected
+
+      # Upsert any non-deleted records
+      # @param records [Array<Array>] List of [key, value] pairs for a group of
+      # non-tombstone records.
+      def upsert_records(records)
+        key_cols = key_columns(records)
+
+        messages = records.map { |k, v| record_attributes(k, v) }
+
+        # If record_attributes indicated no record, skip it
+        messages.compact!
+
+        options = if key_cols.empty?
+                    {} # Can't upsert with no key, just do regular insert
+                  else
+                    {
+                      on_duplicate_key_update: {
+                        # conflict_target must explicitly list the columns for
+                        # Postgres and SQLite
+                        conflict_target: key_cols,
+                        columns: :all
+                      }
+                    }
+                  end
+
+        @klass.import!(messages, options)
+      end
+
+      # Delete any records with a tombstone.
+      # @param records [Array<Array>] List of [key, nil] pairs for a group of
+      # deleted records.
+      def remove_records(records)
+        clause = deleted_query(records)
+
+        clause.delete_all
+      end
+
+      # Get unique key for the ActiveRecord instance from the incoming key.
+      # Override this method (with super) to customize the set of attributes that
+      # uniquely identifies each record in the database.
+      # @param key [Hash] The decoded key.
+      # @return [Hash] The key attributes.
+      def record_key(key)
+        decoded_key = decode_key(key)
+
+        if decoded_key.nil?
+          {}
+        elsif decoded_key.is_a?(Hash)
+          @key_converter.convert(decoded_key)
+        else
+          { @klass.primary_key => decoded_key }
```

Comment:
FYI #21 would still need to be fixed with this. I think it might be too much work to be bundled in here though.